add nerf viewer

WYK96 opened the pull request:
add nerf viewer client and server.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openxrlab/xrprimer?pullRequest=40) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/openxrlab/xrprimer?pullRequest=40) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: weiyukun1<br/>:x: WYK96<br/><hr/>**weiyukun1** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openxrlab/xrprimer?pullRequest=40) it.</sub>

GACLove commented:
1. Add Architecture diagram in README.md
2. Delete the top-level directory "nerf_viewer"
3. Rename the directory "nerf_viewer" to "viewer":
4. Rename the directory "bridge_server" to "server"
5. Write a `start.sh` or `start.py` script to launch the server and client 
6. Generate a default .gitignore file, visit https://www.toptal.com/developers/gitignore.
7. To delete unused files and modules

yl-1993 commented on viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py:
```diff
@@ -0,0 +1,85 @@
+import atexit
```
Shall we rename it to `bridge_server.py`? As there is no conflict, it is not necessary to add `_subprocess` suffix.

WYK96 replied:
1. Renamed `bridge_server_subprocess.py` to `bridge_server.py`
2. Renamed function `run_bridge_server_as_subprocess` to `start_bridge_server`

yl-1993 commented on viewers/nerf_viewer/client/README.md:
```diff
@@ -0,0 +1,70 @@
+
+### `npm run build` fails to minify
+
+This section has moved here: [https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify](https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify)
```
For `Learn More` section, as each subsection. only as one sentence, shall we use bullet point directly?   For example,

- [Code Splitting](https://facebook.github.io/create-react-app/docs/code-splitting)
- [Analyzing the Bundle Size](https://facebook.github.io/create-react-app/docs/analyzing-the-bundle-size)
- ...

WYK96 replied:
This README.md is automatically generated by [Create React App](https://create-react-app.dev/). It has been replaced to [XRNerf Viewer](https://github.com/openxrlab/xrprimer/pull/40/commits/4f7406de9ef05f41e7df00e5398af842012c1da6#diff-46bd9e24f83032c882309d2f34f5ea611782227fcff9e138f59ac955ab3a19a5).

yl-1993 commented on viewers/nerf_viewer/client/src/themes/theme.ts:
```diff
@@ -0,0 +1,180 @@
+    //   },
+    // },
+  },
+});
```
Remove commented content if they are unnecessary anymore.

WYK96 replied:
Unused code blocks removed.

yl-1993 commented:
1. Move two readme files to `${ROOT}/docs/en/visualization/`, e.g., nerf_viewer.md, so it can also be viewed via[readthedocs](https://xrprimer.readthedocs.io/en/latest/).
2. Move all png to `${ROOT}/resources/viewer/` and try to compress the single filesize under 10KB.

Maoxie commented:
+ format *.py files according to the project's `.pre-commit-config.yaml`

LazyBusyYang commented on viewers/nerf_viewer/README.md:
```diff
@@ -0,0 +1,85 @@
+
+Please follow the instructions in:
+
+https://juejin.cn/post/7120267871950733320/
```
Do we have any installation instruction in English? Say https://github.com/nvm-sh/nvm#installing-and-updating.

WYK96 replied:
The original version of README.md is for debug purpose. In the lastest doc, the [installation instructions](https://github.com/WYK96/xrprimer/blob/5b76711b953035e6e1fc3191efe19f66a4fbecf3/docs/en/web_renderer/xrnerf/README.md?plain=1#L27) has been updated so that it is suitable for non-chinese users.

LazyBusyYang commented on viewers/nerf_viewer/README.md:
```diff
@@ -0,0 +1,85 @@
+conda activate BridgeServer
+
+# install dependencies
+pip install -r ./requirements.txt
```
We have a requirements folder in the root directory of the repository, where many different requirements files are named according to their functionality and placed here, such as `requirements/synbody.txt`. You can also consider renaming your `requirements.txt` to something like `bridge_server.txt`. In addition, most of the relative paths in our documentation are relative to the root directory of the repository, not to subdirectories like `viewers`.

WYK96 replied:
For requriements, the dependencies list of server has been moved from 
 `python/xrprimer/services/xrnerf/requirements.txt` to `requirements/service_xrnerf.txt`. 

For paths, they are now relative to the root directory of the repository, e.g., `./run.py` to `xrprimer/python/xrprimer/services/xrnerf/run.py`.

Appreciate for the suggestion.

LazyBusyYang commented on viewers/nerf_viewer/README.md:
````diff
@@ -0,0 +1,85 @@
+#### 2.2 Start Server
+
+```
+python ./run_viewer.py
````
```suggestion
python .python/tools/un_viewer.py
```
We often put files with a main function program entry point in the `tools` folder to distinguish them from files that are only called and not used as entry points.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/.gitignore:
```diff
@@ -0,0 +1 @@
+.idea
```
What if we merge this file with the [.gitignore](https://github.com/openxrlab/xrprimer/blob/main/.gitignore) file in the root directory?

WYK96 replied:
The original .gitignore file is for debug purpose only. It has been removed now.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py:
```diff
@@ -0,0 +1,85 @@
+        return False
+
+
+def get_free_port(default_port: int = None):
```
```suggestion
from typing import Union
def get_free_port(default_port: Union[int, None] = None):
```

WYK96 replied:
The `default_port: int` has been declared as `default_port: Union[int, None]`.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py:
```diff
@@ -0,0 +1,85 @@
+        return False
+
+
+def get_free_port(default_port: int = None):
```
For users, there are two possible scenarios. In the first scenario, the user does not care which port is used as long as the service is started successfully and the port used is returned. The current function can meet this requirement. In the second scenario, the user may be required to use a specific port, such as port 80. If the target port is occupied, an exception should be thrown and there is no need to continue running. Considering these two scenarios, we can modify the parameter list of this function, for example:
1. `def bind_port(allow_random_port: bool, preferred_port: Union[int, None] = None) -> int:`
2. `def bind_port(target_port: Union[int, None] = None) -> int:`

In the first example, the `allow_random_port` parameter is used to determine which usage scenario to use, and `preferred_port` is the port number that is preferentially attempted. In the second example, the `target_port` parameter is used to determine the usage scenario. If it is provided, the target port must be used. If it is not provided, there are no requirements and any available port can be returned.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py:
```diff
@@ -0,0 +1,85 @@
+
+def run_bridge_server_as_subprocess(
+    websocket_port: int,
+    zmq_port: Optional[int] = None,
```
ChatGPT:
In Python, Union[int, None] and Optional[int] are similar in that they both indicate that a variable can be either an integer or None. However, there is a subtle difference between them.

Union[int, None] explicitly specifies that the variable can be either an integer or None. It does not allow any other type to be assigned to it. On the other hand, Optional[int] is actually an alias for Union[int, None], which means it represents the same set of possible types. However, Optional[int] does not explicitly indicate that the variable cannot be assigned any other type.

Therefore, using Union[int, None] is more specific and explicit, and may help prevent type-related bugs in your code. However, both Optional[int] and Union[int, None] can be used interchangeably in most cases.

WYK96 replied:
Since Optional[int] is equivalent to Union[int, None] whereas more concise and readable, we're using Optional as the type annotation.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py:
```diff
@@ -0,0 +1,85 @@
+def run_bridge_server_as_subprocess(
+    websocket_port: int,
+    zmq_port: Optional[int] = None,
+    ip_address: str = "127.0.0.1",
```
`0.0.0.0` may allow access to a larger range than `127.0.0.1`. Should we change the default value to `0.0.0.0`?

WYK96 replied:
Replace the default ip address to 0.0.0.0 makes zmq fail to connect with client.

`TODO: figure out why this happens.`

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/bridge_server_subprocess.py:
```diff
@@ -0,0 +1,85 @@
+            time.sleep(0.5)
+
+        message = "The bridge server subprocess failed."
+        cleanup(process)
```
`process` in `cleanup` is passed by arg list, but in `poll_process` it looks like a global var. 
Considering that the `process` variable has a clear lifetime and these functions are strongly related to it, would it be better to rewrite this code as a class and access it through `self.sub_process`?

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/requirements.txt:
```diff
@@ -0,0 +1,5 @@
+pyzmq==25.0.2
```
Same as https://github.com/openxrlab/xrprimer/pull/40/files#r1191927947

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/viewer_state.py:
```diff
@@ -0,0 +1,92 @@
+import umsgpack
+from typing import Tuple
+import numpy as np
+import cv2 as cv
```
What about not as `cv`? We use `cv2` directly in other modules inside XRPrimer.

WYK96 replied:
Changed `cv` to `cv2`. The OpenCV is used to generate fake images for debug purpose, which should be further replaced by the nerf backend. Hence, the dependency will be removed once the nerf backend is ready.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/viewer_state.py:
```diff
@@ -0,0 +1,92 @@
+        elif str(self.state.resolution) == "1080":
+            return 1920, 1080
+        else:
+            print("invalid resolution")
```
Please check [logger example](https://github.com/openxrlab/xrprimer/blob/main/python/xrprimer/visualization/palette/point_palette.py#L39) in class.
`self.logger.error(msg)` logs the message in error level, if a file handler has been configured, the message will never be lost. After logging, by using raise ValueError, we can prevent the program from continuing to run and causing unexpected errors, which can make it easier for developers to identify the location where the error occurred.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/viewer_state.py:
```diff
@@ -0,0 +1,92 @@
+        Draw an image using current camera configuration.
+        """
+
+        """fetch data from state"""
```
```suggestion
        # fetch data from state
```
pre-commit will find everything breaks rules.

WYK96 replied:
The issue has been fixed by pre-commit.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/viewer_state.py:
```diff
@@ -0,0 +1,92 @@
+        """
+        Let the renderer relief for some time to avoid websocket blocking         
+        """
+        time.sleep(0.05)
```
Why 0.05? It looks like 20 fps. If the time parameter for sleep is a user-specified parameter that is set during construction or dynamically determined based on the program's running state, it may be better.

WYK96 replied:
The relief time is a time step that the backend sleep for some time to avoid websocket blocking. The value of it needs to be further evaluated. A more reasonable relief time would be:

$relief\\_time = time\\_step - (backend\\_inference\\_elapsed + viewer\\_render\\_elapsed)$

e.g. If we want the viewer renders at 60fps, our per-frame budget is 16.6ms. Assuming that the nerf inference time is 5ms and viewer render latency is 4ms, the relif time is 16.6-5-4=7.6ms

Unfortunately, we do not have a ***really*** available backend up to now.

`TODO: replace to a more reasonable relief time once the nerf backend is ready.`

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/visualizer.py:
```diff
@@ -0,0 +1,131 @@
+            thread.start()
+            thread.join(timeout_in_sec)
+        except Exception as e:
+            print("Failed to start thread")
```
Same as https://github.com/openxrlab/xrprimer/pull/40/files#r1191947804.

LazyBusyYang commented on viewers/nerf_viewer/bridge_server/visualizer.py:
```diff
@@ -0,0 +1,131 @@
+        try:
+            thread.start()
+            thread.join(timeout_in_sec)
+        except Exception as e:
```
In most cases, we do not use `Exception` because it corresponds to too broad of an error category. If we can identify a specific type of error, such as `RuntimeError` or `ValueError`, we should only catch the known and specific types of errors.

LazyBusyYang commented on viewers/nerf_viewer/client/.gitignore:
```diff
@@ -0,0 +1,23 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Same as https://github.com/openxrlab/xrprimer/pull/40/files#r1191930152

LazyBusyYang commented on viewers/nerf_viewer/client/.gitignore:
```diff
@@ -0,0 +1,23 @@
+.env.test.local
+.env.production.local
+
+npm-debug.log*
```
Is it possible to redirect these logs to a designated path, such as `logs/npm_log.txt`?

LazyBusyYang commented on viewers/nerf_viewer/client/src/themes/theme.ts:
```diff
@@ -0,0 +1,180 @@
+        },
+      },
+    },
+    // MuiIcon: {
```
If these lines does not work, and it is no important comment, please remove them from official commit. To keep them, push another branch in a forked repo.

WYK96 replied:
Unused code blocks removed.

WYK96 commented on viewers/nerf_viewer/README.md:
````diff
@@ -0,0 +1,85 @@
+pip install -r ./requirements.txt
+```
+
+#### 2.2 Start Server
````
Sorry for the typo. The title hierarchy issue has been fixed.

WYK96 commented on viewers/nerf_viewer/README.md:
```diff
@@ -0,0 +1,85 @@
+
+Visit http://localhost:3000/ in the browser:
+
+![alt viewer](./doc/viewer.png)
```
The original version of REAME.md is for debug purpose. In the lastest doc, the screenshots have been removed.

WYK96 commented on viewers/nerf_viewer/README.md:
```diff
@@ -0,0 +1,85 @@
+# Nerf Viewer Quick Start
```
Glitch fixed: Nerf Viewer -> XRNerf Viewer

WYK96 commented on viewers/nerf_viewer/client/README.md:
```diff
@@ -0,0 +1,70 @@
+
+In the project directory, you can run:
+
+### `npm start`
```
This README.md is automatically generated by [Create React App](https://create-react-app.dev/). It has been replaced to [XRNerf Viewer](https://github.com/openxrlab/xrprimer/pull/40/commits/4f7406de9ef05f41e7df00e5398af842012c1da6#diff-46bd9e24f83032c882309d2f34f5ea611782227fcff9e138f59ac955ab3a19a5).

WYK96 commented:
> 1. Move two readme files to `${ROOT}/docs/en/visualization/`, e.g., nerf_viewer.md, so it can also be viewed via[readthedocs](https://xrprimer.readthedocs.io/en/latest/).
> 2. Move all png to `${ROOT}/resources/viewer/` and try to compress the single filesize under 10KB.

1. Moved client and server readme files to `${ROOT}/docs/en/web_renderer/xrnerf` and `${ROOT}/docs/en/services/xrnerf`, repectively.
2. Moved helper images to `${ROOT}/resources`.  Since the original version of README.md is for debug purpose, the screenshots have been removed in the lastest doc.

WYK96 commented:
> 1. Add Architecture diagram in README.md
> 2. Delete the top-level directory "nerf_viewer"
> 3. Rename the directory "nerf_viewer" to "viewer":
> 4. Rename the directory "bridge_server" to "server"
> 5. Write a `start.sh` or `start.py` script to launch the server and client
> 6. Generate a default .gitignore file, visit https://www.toptal.com/developers/gitignore.
> 7. To delete unused files and modules

1. Added a framework figure in `${ROOT}/docs/en/web_renderer/xrnerf/README.md`
2. The top-level directory `nerf_viewer` has been deleted. Now the viewer client is on `${ROOT}/web_renderer/xrnerf` and the viewer server is on `${ROOT}/python/xrprimer/services/xrnerf`
3. The directory `${ROOT}/viewers/nerf_viewer/client` is renamed to `${ROOT}/web_renderer/xrnerf`
4. The directory `${ROOT}/viewers/nerf_viewer/bridge_server` is renamed to  `${ROOT}/python/xrprimer/services/xrnerf`
5. Added `${ROOT}/web_renderer/xrnerf/start.bat` to launch the client and  `${ROOT}/python/xrprimer/services/xrnerf/run.py` to launch the server.
`TODO: test the framework on Linux systems and add a start.sh for client startup`
6. `TODO: it seems that https://www.toptal.com/developers/gitignore does not provide an option to generate JavaScript and TypeScript .gitignore file`. Need to be fixed later.
7. Deleted redundant files and modules:
   - Removed redundant redux stored variables
   - Removed unused code blocks

WYK96 commented:
> * format `*.py` files via pre-commit hooks according to the project's `.pre-commit-config.yaml`
> * add one blank new line to the end of code files if there is not one.

1. Reformatted python scripts with `${ROOT}/.pre-commit-config.yaml` utilizing pre-commit hooks
2. For each python script, added a new line at the end of file

WYK96 commented on viewers/nerf_viewer/README.md:
```diff
@@ -0,0 +1,85 @@
+
+### 1.1 Install Dependencies
+
+#### 1.1.1 Install NVM
```
The original version of README.md is for debug purpose. In the lastest doc, the installation guide has been replaced to:

The `package.json` contains dependencies to build the viewer, which can be installed using Node Package Manager(`npm`). It's suggested to manage `Node.js` installations with Node Version Manager(`nvm`). The installation instructions of `nvm` can be found [here](https://heynode.com/tutorial/install-nodejs-locally-nvm/).

WYK96 commented on viewers/nerf_viewer/README.md:
```diff
@@ -0,0 +1,85 @@
+
+Please follow the instructions in:
+
+https://juejin.cn/post/7120267871950733320/
```
The original version of README.md is for debug purpose. In the lastest doc, the link for nvm installation guidelines has been replaced from https://juejin.cn/post/7120267871950733320/ to https://heynode.com/tutorial/install-nodejs-locally-nvm/.

WYK96 commented on viewers/nerf_viewer/README.md:
````diff
@@ -0,0 +1,85 @@
+
+```shell
+# install node 18.15.0
+nvm install 18.15.0
````
Patch version has been removed: `18.15.0` -> `18.15`

WYK96 commented on viewers/nerf_viewer/bridge_server/server.py:
```diff
@@ -0,0 +1,173 @@
+        # type_ = m["type"]
+        # path = list(filter(lambda x: len(x) > 0, m["path"].split("/")))
+
+        if type == UPDATE_CAMERA_TRANSLATION:
```
Wrapped actions into enums.

WYK96 commented on viewers/nerf_viewer/bridge_server/actions.py:
```diff
@@ -0,0 +1,23 @@
+# update the scene state
+UPDATE_STATE = "UPDATE_STATE"
+
+UPDATE_RENDER_RESULT = "UPDATE_RENDER_RESULT"
```
Wrapped actions into enums.

WYK96 commented on viewers/nerf_viewer/bridge_server/server.py:
```diff
@@ -0,0 +1,173 @@
+        websocket_to_skip: Optional[WebSocketHandler] = None
+    ):
+        """forward a zmq message to all websockets"""
+        """nerf backend -> viewer"""
```
The comment `"""nerf backend -> viewer"""` is no longer useful and has been deleted.

WYK96 commented on viewers/nerf_viewer/bridge_server/server.py:
```diff
@@ -0,0 +1,173 @@
+        websocket_port: int = 4567,
+        ip_address: str = "127.0.0.1",
+        use_ngrok: bool = False
+):
```
Extra indents has been fixed by pre-commit.

WYK96 commented on viewers/nerf_viewer/bridge_server/viewer_state.py:
```diff
@@ -0,0 +1,92 @@
+import cv2 as cv
+from actions import UPDATE_STATE, UPDATE_RENDER_RESULT
+from bridge_server_subprocess import run_bridge_server_as_subprocess
+from visualizer import Viewer
```
All python imports have been sorted by isort in pre-commit.

WYK96 commented on viewers/nerf_viewer/client/src/App.css:
```diff
@@ -0,0 +1,38 @@
+/* .App {
```
Unused file removed.

WYK96 commented on viewers/nerf_viewer/client/src/SceneNode.js:
```diff
@@ -0,0 +1,31 @@
+// /* eslint-disable no-underscore-dangle */
```
Unused file removed.
